bindings/java/c/host.c

```diff
@@ -117,159 +117,155 @@ static enum evmc_storage_status set_storage_fn(struct evmc_host_context* context
-
-        // set java method params
-        jcontext_index = context->index;
-        jaddress = CopyDataToJava(jenv, address, sizeof(struct evmc_address));
-        jkey = CopyDataToJava(jenv, key, sizeof(struct evmc_bytes32));
-        jval = CopyDataToJava(jenv, value, sizeof(struct evmc_bytes32));
-
-        // call java method
-        jint jresult = (*jenv)->CallStaticIntMethod(jenv, host_class, method, jcontext_index,
-                                                    jaddress, jkey, jval);
-        result = (enum evmc_storage_status)jresult;
-    }
+
+    jclass host_class;
+    jmethodID method;
+    jint jcontext_index;
+    jbyteArray jaddress;
+    jbyteArray jkey;
+    jbyteArray jval;
+
+    // get java class
+    host_class = (*jenv)->FindClass(jenv, "org/ethereum/evmc/Host");
+    assert(host_class != NULL);
+
+    // get java method
+    method = (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
+    assert(method != NULL);
+
+    // set java method params
+    jcontext_index = context->index;
+    jaddress = CopyDataToJava(jenv, address, sizeof(struct evmc_address));
+    jkey = CopyDataToJava(jenv, key, sizeof(struct evmc_bytes32));
+    jval = CopyDataToJava(jenv, value, sizeof(struct evmc_bytes32));
+
+    // call java method
+    jint jresult = (*jenv)->CallStaticIntMethod(jenv, host_class, method, jcontext_index, jaddress,
+                                                jkey, jval);
+    result = (enum evmc_storage_status)jresult;
+
     return result;
 }
 
 static evmc_uint256be get_balance_fn(struct evmc_host_context* context, const evmc_address* address)
 {
+#ifdef DEBUG
+    printf("Entering get_balance_fn\n");
+#endif
     evmc_uint256be result;
     char java_method_name[] = "get_balance";
     char java_method_signature[] = "(I[B)Ljava/nio/ByteBuffer;";
     assert(context != NULL);
     JNIEnv* jenv = attach();
-    if (jenv != NULL)
-    {
-        jclass host_class;
-        jmethodID method;
-        jint jcontext_index;
-        jbyteArray jaddress;
-
-        // get java class
-        host_class = (*jenv)->FindClass(jenv, "org/ethereum/evmc/Host");
-        assert(host_class != NULL);
-
-        // get java method
-        method =
-            (*jenv)->GetStaticMethodID(jenv, host_class, java_method_name, java_method_signature);
-        assert(method != NULL);
-
-        // set java method params
-        jcontext_index = context->index;
-        jaddress = CopyDataToJava(jenv, address, sizeof(struct evmc_address));
-
-        // call java method
-        jobject jresult =
-            (*jenv)->CallStaticObjectMethod(jenv, host_class, method, jcontext_index, jaddress);
-        assert(jresult != NULL);
-
-        evmc_uint256be* result_ptr =
-            (evmc_uint256be*)(*jenv)->GetDirectBufferAddress(jenv, jresult);
-        assert(result_ptr != NULL);
-        result = *result_ptr;
-
-        (*jenv)->ReleaseByteArrayElements(jenv, jaddress, (jbyte*)address, 0);
```

Comment:
This was segfaulting when running against vm reference tests. Removed that line.

Reply:
We should open an issue to develop a more complete set of unit tests to catch this.  

If this causes a segfault, then I would think these functions would have the same problem:

in `get_code_hash_fn()`:
https://github.com/ethereum/evmc/blob/0c9eb8e30b24e18e2910aef98c8a84c33d1360a8/bindings/java/c/host.c#L268

in `copy_code_fn()`:
https://github.com/ethereum/evmc/blob/0c9eb8e30b24e18e2910aef98c8a84c33d1360a8/bindings/java/c/host.c#L317

Reply:
Yeah, neither fail. The reference tests should cover them.

Hard to tell for me why it segfaults. The JVM exits with error code 134, which means exit code 6 + 128.
Exit code 6 is a SIGABRT.